syncer/syncer.go

```diff
@@ -1803,11 +1822,27 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 			}
 
 			if s.streamerController.CanRetry(err) {
-				err = s.streamerController.ResetReplicationSyncer(tctx, lastLocation)
+				// GlobalPoint is the last finished GTID
+				err = s.streamerController.ResetReplicationSyncer(tctx, s.checkpoint.GlobalPoint())
 				if err != nil {
 					return err
 				}
-				log.L().Info("reset replication binlog puller")
+				log.L().Info("reset replication binlog puller", zap.Any("pos", s.checkpoint.GlobalPoint()))
+				if s.cfg.EnableGTID {
+					for i := 0; i < eventIndex; {
+						e, err = s.getEvent(tctx, currentLocation)
+						if err != nil {
+							return err
+						}
+						if _, ok := e.Event.(*replication.RowsEvent); ok {
+							i++
```

Comment:
pelase tell me why we can drop the first n events here? 
will this result in data loss from last saved checkpoint until the `eventIndex` event

Reply:
i think we should alway into safemode for one trx in this situation

Reply:
In GTID mode, the reader will start for the first event of the transaction after the `streamerController` is reset. So we should skip the first N rowsEevets because they have been processed. Though enter safe mode is another way to do so, but just skip should be a more straight forward way to avoid duplication. Maybe you need to take a deeper look at the changes.😅

Reply:
> first N rowsEevets because they have been processed

i don't think our code ensure this, for example one tx have 3 events  and getEvent met error at the second event,
in this case, all we can guarantee is that event1's sql is already in the queue, not that it has been executed downstream, i am sure if this will result in data loss🤔, so i perfer to alway into safemode. in addition the connection-level retry already serves our purpose of not interrupting the task

but ignore first N event in relay unit is ok to me , because we can make sure all event is write to file

Reply:
I don't get the point in what condition in your example will it cause a data loss, can you explain it in detail @Ehco1996

Reply:
https://github.com/pingcap/dm/pull/2047/files#r699233692 I think in this comment, @lance6716 also prefer to just skip the duplicate event instead of relay on safe-mode.

Reply:
The former logic also directly retry some error instead of enter Pause&Resume cycle when meets some error in GetEvent, this PR just resolve the issue that the former logic may cause duplicate error when GTID is enabled.😅